### ISSUE TITLE
[v1.0-dev] Handle click on embed object

### ISSUE THREAD

**Issue by hnvn** [memspace/zefyr#426]
First of all, many thanks for this great package. I am using v1.0-dev, it comes with more ability of customization embed objects and it works well in aspect of rendering object. But I face a problem when I try to handle click event on my embed object, such as: click on Play button on my object. The editor always captures the click event and make the object to be selected state, click events don't hit my button at all. How to get my button work in this case? Thanks in advance.

**Comment by cgestes** [comment]
Hi,

this is known and hard to fix. @pulyaevskiy is aware of it, and will work on it.

I did the following on the previous version to make it work:
(this was a quick hack, but it may help to get some inspiration. the commits are not clean...it was on a dirty branch)
https://github.com/cgestes/zefyr/commit/667328e357cc5d00030035e73af309d3a0520be6
There may be part of the functionality in here also: https://github.com/cgestes/zefyr/commit/53602733fe965fb8eff0824a9212ab1fd10d433a

**Comment by pulyaevskiy** [comment]
Yep, @cgestes is correct. This still needs to be addressed. I don't have any updates on this yet, but this is high on my todo list.

**Comment by hnvn** [comment]
Hi @pulyaevskiy , do you have plan to work on this issue in next few weeks? thanks in advance

**Comment by pratikpparikh** [comment]
@pulyaevskiy @cgestes can you please provide some direction so we can add the support.

**Comment by cgestes** [comment]
Hi @pratikpparikh 

The main idea is to tweak hit testing so that the events goes to the embedded widgets instead of the selection layer.

I don't know exactly what has to be done, the quick hack I gave above will work.

I wonder what should we do if we want to allow selection to work on an embedded object. There lay be some other corner cases.

**Comment by pratikpparikh** [comment]
Thank you @cgestes ,  I tried many options and non delivered the correct delegation of responsibility for handling click to object including TextSpan.  I know think it is a problem due to rendering view being stateless.   @pulyaevskiy could this be a cause of new architecture?  were the view is now stateless?

**Comment by nipunasudha** [comment]
Hope you all are safe during the pandemic. Any progress on this please? Looking forward to using some interactive embeds.

**Comment by nipunasudha** [comment]
I think we can at least enable click events in readonly view? Wouldn't that be easier because there is no cursor competing for gestures?

**Comment by Amir-P** [comment]
> I think we can at least enable click events in readonly view? Wouldn't that be easier because there is no cursor competing for gestures?

I don't think that would be possible. Caret's are being displayed and interactions are handled even in read-only mode for selection and copying. @nipunasudha

**Comment by nipunasudha** [comment]
To anyone who go through this thread, I ended up Using `Overlay()` widget to render overlayed content for the things that needs to be interactive. And it works really well!

**Comment by nipunasudha** [comment]
Like I said I implemented it using `Overlay` widgets. For fixed size content this works amazingly well. But my embedded content has dynamic heights which depends on the the overlay itself! 😆 I'm trying my hardest but I can't figure out a stable way of sizing the Zefyr embedded widget according to the dynamic overlay height. 

If someone has any idea to how to achieve that without hacky messaging channels between build() calls, please tell me how. And I'm gonna let this go for now because it's on the 1.0 milestone to fix this issue.

**Comment by cgestes** [comment]
I dont remember the details, I will have another look at it, but this is
what I implemented to have click transmitted to embedded widget:
https://github.com/cgestes/zefyr/commit/667328e357cc5d00030035e73af309d3a0520be6

It probably goes with a few commits around it. I still didn't take the time
to clean it. But It was working.

**Comment by cgestes** [comment]
@pulyaevskiy any idea how to fix this todo?
https://github.com/memspace/zefyr/blob/e7433f78c2362d569d0c1af702543432efdf263d/packages/zefyr/lib/src/widgets/editor.dart#L394
This would help with changing the mouseCursor on links also. (We can specify them in the textspan)

**Comment by cgestes** [comment]
@nipunasudha I suspect this is the proper fix: https://github.com/cgestes/zefyr/tree/hit-test-children

It is still work in progress and have some offset issues here and there, but would be happy to have feedback if it fixes the issue you had.

**Comment by nipunasudha** [comment]
> @nipunasudha I suspect this is the proper fix: https://github.com/cgestes/zefyr/tree/hit-test-children
> 
> It is still work in progress and have some offset issues here and there, but would be happy to have feedback if it fixes the issue you had.

Amazing! I'll test the branch within a couple of days and provide you feedback. Thank you so much for your hard work.

**Comment by cgestes** [comment]
this is most probably fixed by #543  (same branch, just linking the issue).

Happily waiting for your feedback on the subject :)

**Comment by nipunasudha** [comment]
> this is most probably fixed by #543 (same branch, just linking the issue).
> 
> Happily waiting for your feedback on the subject :)

Migrated to zefyr, ported some zefyrka attributes into my fork. Now testing your PR. Will report back soon.

**Comment by nipunasudha** [comment]
> this is most probably fixed by #543 (same branch, just linking the issue).
> 
> Happily waiting for your feedback on the subject :)

I tested it out and it works really well with 'text lines'. The interaction works well! Here is the experiment I did.

1. I modified the `defaultZefyrEmbedBuilder` to add a gesture detector at the end of the `hr` like this.

```dart
  if (node.value.type == 'hr') {
    final theme = ZefyrTheme.of(context)!;
    return Row(
      children: [
        Expanded(
          child: Divider(
            height: theme.paragraph.style.fontSize! * theme.paragraph.style.height!,
            thickness: 2,
            color: Colors.grey.shade200,
          ),
        ),
        GestureDetector( // new widget
          child: const Icon(Icons.extension),
          onLongPress: () => print('Item long pressed!!'),
          onTap: () => print('Item clicked!'),
        ),
      ],
    );
  }

```
And both callbacks were fired correctly. Only if the long press callback given, default long press action on the `hr` is not fired, which is the desirable behavior.

2. Then I tried horizontal & vertical scroll views inside the embedded object. And I'm over the moon!

https://user-images.githubusercontent.com/17500323/140718226-c0ccd899-b378-4af9-8daf-011f20c37f70.mov

These are my findings, and everything seems to work really well. I'm planning on finding out how well this works with in-app webviews. Please ask me if you need anything else tested for this.

**Comment by nipunasudha** [comment]
https://user-images.githubusercontent.com/17500323/141107130-172b7d04-6b13-4c51-b425-30d402dba81a.mov

**Comment by cgestes** [comment]
Great. 👍. We will merge that for 1.0.
A lot of thanks goes to @pulyaevskiy for doing all the hard job of putting the architecture of the editor together.

**Comment by pulyaevskiy** [comment]
The hit test fix has been merged into 1.0-dev. Thanks @cgestes for finding and fixing it and @nipunasudha for verifying it works on your end.

Considering this done.
